stdlib/InteractiveUtils/src/macros.jl

```diff
@@ -206,15 +211,167 @@ function merge_namedtuple_types(nt::Type{<:NamedTuple}, nts::Type{<:NamedTuple}.
             end
         end
     end
-    NamedTuple{Tuple(names), Tuple{types...}}
+    return NamedTuple{Tuple(names), Tuple{types...}}
+end
+
+function gen_call(fcn, args, where_params, kws; use_signature_tuple::Bool, not_an_opaque_closure::Bool = true)
+    f, args... = args
+    args = collect(Any, args)
+    if !use_signature_tuple
+        f = esc(reescape(extract_farg, f))
+        tt = typesof_expr(args, where_params)
+        return :($fcn($f, $tt; $(kws...)))
+    end
+    # We use a signature tuple only if we are sure we won't get an opaque closure as first argument.
+    # If we do get one, we have to use the 2-argument form.
+    if isexpr(f, :(::)) || not_an_opaque_closure
+        # We have a type, not a value, so not an opaque closure.
+        sigt = typesof_expr(Any[f, args...], where_params)
+        return :($fcn($sigt; $(kws...)))
+    end
+    tt = typesof_expr(args, where_params)
+    sigt = typesof_expr_unescaped(Any[:f, esc.(args)...], where_params)
+    return quote
+        f = $(esc(f))
+        if isa(f, Core.OpaqueClosure)
+            $fcn(f, $tt; $(kws...))
+        else
+            $fcn($sigt; $(kws...))
+        end
+    end
+end
+
+function expand_ref_begin_end!(f::Function, ex, __module__::Module)
+    arr = ex.args[1]
+    args = copy(ex.args)
+    new = replace_ref_begin_end!(__module__, ex)
+    modified = ex.args .≠ args
+    if any(modified) && (isexpr(arr, :(::), 1) || isexpr(arr, :(::), 2) || isexpr(arr, :..., 1))
+        return Expr(:call, :error, "`begin` or `end` cannot be used with a type-annotated left-hand side argument for an indexing syntax")
+    end
+    call = f(ex)
+    !any(modified) && return call
+    fixup_hygiene_for_ref_temporary!(new)
+    # We have to mutate `ex`, then return `new` which evaluates `arr` before use.
+    ex.head = call.head
+    ex.args = call.args
+    return new
+end
+
+function fixup_hygiene_for_ref_temporary!(ex)
+    # Match the local variable `##S#...` so we may escape its definition.
+    # We don't want to use `escs = 1` in `replace_ref_begin_end_!` because
+    # then we delegate escaping to this function, whereas we otherwise manage
+    # ourselves the escaping in all other code paths.
+    isexpr(ex, :block) || return
+    decl = ex.args[1]
+    isexpr(decl, :local, 1) || return
+    assignment = decl.args[1]
+    isexpr(assignment, :(=), 2) || return
+    variable = assignment.args[1]
+    startswith(string(variable), "##S#") || return
```

Comment:
This seems like a bad idea to find some random variable by name and then apply random mutation to it (mainly it seems it would result in incorrect code being executed if someone has a statement with multiple `begin` or `end` like `a[end][end]`)